DGAN event sequences

mvansegbroeck opened the pull request:
Adding notebook and helper functions in support of the DGAN blogpost

johnnygreco commented on docs/notebooks/demo/gretel-demo-sequence-of-events.ipynb:
```diff
@@ -0,0 +1,456 @@
+    "max_len = df.groupby(EXAMPLE_COLUMN).size().max()\n",
+    "\n",
+    "# Pad each group and concatenate back into a DataFrame\n",
+    "data_source = pd.concat(\n",
+    "    [pad_sequence(group, max_len, example_id_column=EXAMPLE_COLUMN, event_column=EVENT_COLUMN) for _, group in df.groupby(EXAMPLE_COLUMN)],\n",
+    "    ignore_index=True,\n",
```
Might be helpful to add `pad_value="[END]"` as an explicit keyword argument to `pad_sequence` here to help the user understand where the `[END]` value came from.

johnnygreco commented on docs/notebooks/demo/gretel-demo-sequence-of-events.ipynb:
```diff
@@ -0,0 +1,456 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from utils import EventTypeHistogramAndTransitionDistance\n",
```
`utils` -> `demo_utils`

johnnygreco commented on docs/notebooks/demo/gretel-demo-sequence-of-events.ipynb:
```diff
@@ -0,0 +1,456 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import sys\n",
+    "\n",
+    "# show best config \n",
+    "best_config = tuner_results.best_config\n",
+    "yaml.dump(config, sys.stdout, default_flow_style=False, sort_keys=False)"
```
Should you be showing `best_config` rather than `config`?

Also, I think this is simpler and will work

```python 
print(yaml.dump(best_config))
```

Then you can remove the `sys` import.

johnnygreco reviewed (commented):
Nice work, @mvansegbroeck – this is such a useful addition to our demos!

I left comments in the notebook (maybe not the best via github 🤦‍♂️)

A couple high-level questions

- Looks like the plan is to put demo utility modules into `demo-utilities-lib`. In this case, should `demo_utils.py` be renamed to something specific to this demo? 

- Should there be NaNs in the dataset before (or after) padding? I got a NaN in the sample output at the top of the notebook.

<img width="916" alt="Screenshot 2024-02-28 at 10 57 32 AM" src="https://github.com/gretelai/gretel-blueprints/assets/10998105/472053bd-9d40-4855-9fa0-ccffe1d6c9a6">

mvansegbroeck commented on docs/notebooks/demo/gretel-demo-sequence-of-events.ipynb:
```diff
@@ -0,0 +1,456 @@
+    "This dataset describes a project management process. The process is structured into five main phases, with each phase comprising mandatory events and optional events to provide flexibility and adaptability to the project's needs. Here's a breakdown of the dataset based on this workflow:\n",
+    "\n",
+    "### Phases:\n",
+    "1. **Initiation**: The starting point of the project, focused on establishing the project's foundation.\n",
+    "2. **Planning**: Involves detailed preparation and strategizing for how the project will be executed.\n",
+    "3. **Execution**: The phase where the planned activities are carried out to create the project's deliverables.\n",
+    "4. **Monitoring and Controlling**: Concurrent with execution, this phase ensures the project stays on track and adheres to quality standards.\n",
+    "5. **Closure**: Concludes the project by ensuring all aspects are completed satisfactorily and formally closing the project.\n",
```
Will do.

mvansegbroeck commented on docs/notebooks/demo/gretel-demo-sequence-of-events.ipynb:
```diff
@@ -0,0 +1,456 @@
+    "from IPython.display import Image\n",
+    "\n",
+    "# Display an image with its filename\n",
+    "Image(filename=\"project_management_workflow.png\", width=400, height=200)"
```
yes, thanks for that. I was planning to bring the figure into the MD cell above

mvansegbroeck commented:
Thanks Johnny. I believe I have addressed all your comments.
I have renamed demo_utils.py into dgan_tuner_utils.py.
Also took another look at the dataset and resolved the NaN fields there before padding.

johnnygreco reviewed (approved):
Looks great! 

I verified everything runs 🚀
